### ISSUE TITLE
Add a `--locked` flag for `forc` commands that read from the `Forc.lock` file

### ISSUE THREAD

**Issue by mitchmindtree** [FuelLabs/sway#1411]
Very much inspired by [`cargo`'s `--locked` (or `--frozen`) flag](https://doc.rust-lang.org/cargo/commands/cargo-build.html#manifest-options).

Setting this flag would require that the project's `Forc.lock` file is valid for the current state of the `Forc.toml` manifest and would cause an error instead of automatically recreating the lock file.

This is particularly beneficial for CI/testing to ensure that the main branch always has an up-to-date lock file.

**Comment by mitchmindtree** [comment]
#1410 is an motivating example.

**Comment by ra0x3** [comment]
@mitchmindtree Mind if I pick this one up?

**Comment by mitchmindtree** [comment]
Sounds good!

**Comment by ra0x3** [comment]
@mitchmindtree I took a stab at this one earlier today with no luck. What I'm seeing with this is:

1. Add the `--locked` opt to the proper commands (easy enough)
2. Check that the dependencies in `Forc.toml` match what is in `Forc.lock` if the `--locked` flag is passed, and abort if these two do not match
    - I believe this is the piece I have a question about
    - Looking through the code I get the feeling we already have abstractions to do this, however, I can't seem to find it
    - I see in `forc_build.rs` we already have the `manifest: ManifestFile` and the `lock: Lock`. So I'm assuming we just want a way to check that `manifest` vs the `lock`?

Also let me know if I'm totally thinking about this the wrong way

CC @JoshuaBatty

**Comment by mitchmindtree** [comment]
@ra0x3 the main behaviour we are concerned about is here:

https://github.com/FuelLabs/sway/blob/f0beb377018cf6ac16910a0ff2000f202852bffb/forc/src/ops/forc_build.rs#L59

Currently, if the `BuildPlan` that we reconstructed from the `Forc.lock` file fails validation against the `Forc.toml` manifest, or if it fails to load because the lock file simply doesn't exist yet, we automatically create a *new* build plan and re-write the lock file accordingly. This is what is happening in the `or_else` block at the above link.

We want to change this behaviour so that:

1. If `--locked` **is** specified, we fail with an error stating that `--locked` was specified, but the `Forc.lock` file is not up-to-date for the current manifest.
2. If `--locked` is **not** specified, we automatically recreate the build plan and lock file (as we currently do).

In other words, we want to add a condition at the beginning of that `or_else` block that looks something like this:

```rust
    let plan: pkg::BuildPlan = plan_result.or_else(|e| -> Result<pkg::BuildPlan> {
        if locked {
            bail!("`--locked` was specified, but the `Forc.lock` file is not valid for the project's manifest: {}\nTo resolve this, you can update the `Forc.lock` file using the `forc update` command.", e);
        }
        // Everything else stays the same
```
